File: lab1/main.py
import threading
import random
import time
from bank import Bank


def transfer_worker(bank, num_accounts, num_operations, min_amount, max_amount):
    for i in range(num_operations):
        from_id = random.randint(0, num_accounts - 1)
        to_id = random.randint(0, num_accounts - 1)
        while from_id == to_id:
            to_id = random.randint(0, num_accounts - 1)
        
        amount = random.randint(min_amount, max_amount)
        bank.transfer(from_id, to_id, amount)


def consistency_checker(bank, num_checks, check_interval_ms):
    for i in range(num_checks):
        time.sleep(check_interval_ms / 1000.0)
        if bank.consistency_check():
            print(f"Consistency check passed. Total: {bank.get_total_balance()}")
        else:
            print("WARNING: Consistency check failed!")


def main():
    NUM_ACCOUNTS = 10
    INITIAL_BALANCE = 1000
    NUM_OPERATIONS = 20000
    MIN_AMOUNT = 1
    MAX_AMOUNT = 100
    NUM_TRANSFER_THREADS = 4
    NUM_CONSISTENCY_CHECKS = 10
    CHECK_INTERVAL_MS = 200
    
    bank = Bank(NUM_ACCOUNTS, INITIAL_BALANCE)
    
    print("Initial bank state:")
    print(f"Total balance: {bank.get_total_balance()}")
    print(f"Expected total: {bank.get_initial_total()}")
    print()
    
    assert bank.consistency_check()
    
    threads = []
    for i in range(NUM_TRANSFER_THREADS):
        thread = threading.Thread(
            target=transfer_worker,
            args=(bank, NUM_ACCOUNTS, NUM_OPERATIONS, MIN_AMOUNT, MAX_AMOUNT)
        )
        thread.start()
        threads.append(thread)
    
    checker_thread = threading.Thread(
        target=consistency_checker,
        args=(bank, NUM_CONSISTENCY_CHECKS, CHECK_INTERVAL_MS)
    )
    checker_thread.start()
    
    for thread in threads:
        thread.join()
    
    checker_thread.join()
    
    print()
    print("Final bank state:")
    print(f"Total balance: {bank.get_total_balance()}")
    print(f"Expected total: {bank.get_initial_total()}")
    
    if bank.consistency_check():
        print("SUCCESS: Final consistency check passed!")
    else:
        print("FAILURE: Final consistency check failed!")
    
    print()
    print("Individual account balances:")
    for i in range(NUM_ACCOUNTS):
        print(f"Account {i}: {bank.get_account_balance(i)}")


if __name__ == "__main__":
    main()

File: lab1/cpp/main.cpp
#include "bank.h"
#include <thread>
#include <vector>
#include <random>
#include <iostream>
#include <chrono>
#include <cassert>

// Thread function to perform random transfers
void transfer_worker(Bank& bank, int num_accounts, int num_operations, int min_amount, int max_amount) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> account_dist(0, num_accounts - 1);
    std::uniform_int_distribution<> amount_dist(min_amount, max_amount);
    
    for (int i = 0; i < num_operations; i++) {
        int from = account_dist(gen);
        int to = account_dist(gen);
        
        // Make sure from and to are different
        while (from == to) {
            to = account_dist(gen);
        }
        
        int amount = amount_dist(gen);
        bank.transfer(from, to, amount);
    }
}

// Thread function to perform consistency checks
void consistency_checker(Bank& bank, int num_checks, int check_interval_ms) {
    for (int i = 0; i < num_checks; i++) {
        std::this_thread::sleep_for(std::chrono::milliseconds(check_interval_ms));
        
        bool consistent = bank.consistency_check();
        if (!consistent) {
            std::cout << "WARNING: Consistency check failed!" << std::endl;
        } else {
            std::cout << "Consistency check passed. Total: " << bank.get_total_balance() << std::endl;
        }
    }
}

int main() {
    const int NUM_ACCOUNTS = 10;
    const int INITIAL_BALANCE = 1000;
    const int NUM_OPERATIONS = 10000;
    const int MIN_AMOUNT = 1;
    const int MAX_AMOUNT = 100;
    const int NUM_TRANSFER_THREADS = 4;
    const int NUM_CONSISTENCY_CHECKS = 5;
    const int CHECK_INTERVAL_MS = 200;
    
    // Create bank with initial state
    Bank bank(NUM_ACCOUNTS, INITIAL_BALANCE);
    
    std::cout << "Initial bank state:" << std::endl;
    std::cout << "Total balance: " << bank.get_total_balance() << std::endl;
    std::cout << "Expected total: " << bank.get_initial_total() << std::endl;
    std::cout << std::endl;
    
    // Verify initial consistency
    assert(bank.consistency_check());
    
    // Create transfer worker threads
    std::vector<std::thread> threads;
    
    // Start transfer worker threads
    for (int i = 0; i < NUM_TRANSFER_THREADS; i++) {
        threads.emplace_back(transfer_worker, std::ref(bank), NUM_ACCOUNTS, 
                            NUM_OPERATIONS, MIN_AMOUNT, MAX_AMOUNT);
    }
    
    // Start consistency checker thread
    std::thread checker_thread(consistency_checker, std::ref(bank), 
                              NUM_CONSISTENCY_CHECKS, CHECK_INTERVAL_MS);
    
    // Wait for all transfer threads to complete
    for (auto& thread : threads) {
        thread.join();
    }
    
    // Wait for consistency checker thread
    checker_thread.join();
    
    std::cout << std::endl;
    std::cout << "Final bank state:" << std::endl;
    std::cout << "Total balance: " << bank.get_total_balance() << std::endl;
    std::cout << "Expected total: " << bank.get_initial_total() << std::endl;
    
    // Final consistency check
    bool final_consistent = bank.consistency_check();
    
    if (final_consistent) {
        std::cout << "SUCCESS: Final consistency check passed!" << std::endl;
    } else {
        std::cout << "FAILURE: Final consistency check failed!" << std::endl;
    }
    
    // Print individual account balances
    std::cout << std::endl;
    std::cout << "Individual account balances:" << std::endl;
    for (int i = 0; i < NUM_ACCOUNTS; i++) {
        std::cout << "Account " << i << ": " << bank.get_account_balance(i) << std::endl;
    }
    
    return final_consistent ? 0 : 1;
}

